google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/SeekableByteChannelPrefetcher.java

```diff
@@ -0,0 +1,494 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage.contrib.nio;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.NonWritableChannelException;
+import java.nio.channels.SeekableByteChannel;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UnknownFormatConversionException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+
+/**
+ * SeekableByteChannelPrefetcher wraps an existing SeekableByteChannel to add prefetching.
+ * The prefetching is done on a different thread, so you can write simple code that repeatedly
+ * calls read() to get data, processes it, and then calls read again -- and yet this
+ * simple code overlaps computation and communication for you.
+ * (Of course this is only worthwhile if the underlying SeekableByteChannel doesn't already
+ * implement prefetching).
+ */
+public final class SeekableByteChannelPrefetcher implements SeekableByteChannel {
```

Comment:
I forgot to say earlier. I think we should add `@BetaApi` annotation to this.

Reply:
The whole NIO is already marked as beta in the documentation so this seems overkill.